api/external.go

```diff
@@ -35,15 +38,26 @@ type ExternalSignupParams struct {
 	Code     string `json:"code"`
 }
 
+type TokenVerifierParams struct {
+	CodeChallenge string `json:"code_challenge"`
+	CodeVerifier  string `json:"code_verifier"`
+}
+
 // ExternalProviderRedirect redirects the request to the corresponding oauth provider
 func (a *API) ExternalProviderRedirect(w http.ResponseWriter, r *http.Request) error {
+	const codeChallenge = "code_challenge"
 	ctx := r.Context()
 	db := a.db.WithContext(ctx)
 	config := a.config
 
 	query := r.URL.Query()
 	providerType := query.Get("provider")
 	scopes := query.Get("scopes")
+	challenge := query.Get(codeChallenge)
+	err := storage.StoreInSession(codeChallenge, challenge, r, w)
```

Comment:
Let's not use this supposed "session storage." It's best if we store the authorization request in the database instead of cookies.